Merging the merging branch with gps weather data to breathe app

SarahPanther opened the pull request:
MVP 2 and MVP 1 final feature integration into the main code-base for the app.

Youssef-Beltagy committed:
ClimaCell API calls

I implemented the logic for making HTTP requests to get weather information from the ClimaCell API.

I also wrote unit tests to test the feature in Seattle, Spokane, Miami, and a random location in the US.

I need to write test cases to test retrieving historical information and clean up the code.

I have yet to connect this logic with a Broadcast receiver. The next steps would be to either implement the logic that gets the pin data or implement a Broadcast receiver and test the logic for the Inhaler electronics.

SarahPanther committed:
Added some test code to get GPS in the Main Activity - will move to WorkManger but tried it here for now.

SarahPanther committed:
Moved GPS to WeatherDataWorker.

SarahPanther committed:
Merged Climacell branch into this branch

SarahPanther committed:
Upgraded Android Studio to 4.2 and gradle to 4.2

SarahPanther committed:
Display WeatherData info to Main Activity UI (gathered from the WeatherAPIWorker).  This code is quite messy and needs to be re-organized

SarahPanther committed:
Getting GPS and displaying ClimaCell weather data to the screen independently work but cannot pass GPS to the ClimaCell task.  This is because the GPS returns the location after the GPS task completes (works on an @onComplete), so the GPS task consisently fails. Will need to redo the GPS so getting the location happens synchronously and doesn't return from the GPS task until the location is received.

SarahPanther committed:
ListenableWorker GPSWorker now gets GPS; still working for the ListenableWorker to return success/failure

SarahPanther committed:
Used fused location provider; now GPSWorker returns lat/long to WeatherAPIWorker using WorkManager

SarahPanther committed:
Update TaskObjectSerializationHelper.java

got rid of useless import

SarahPanther committed:
Corrected comment

SarahPanther committed:
Weather API call failed sporadically because JSON object returned had some null values. Added some comments to change JSONObject.getDouble() etc. to JSONObject.OptDouble() etc. to get null back instead throwing exception. Added some log statements to exception cases.

SarahPanther committed:
Merged GPS/ClimaCell feature with main codebase.  Will need to still do some major refactoring before the pull request to breathe-app

SarahPanther committed:
Put weather data collection classes in its own package

SarahPanther committed:
Took API call logic out of WeatherData and organized in CollectWeatherData class

SarahPanther committed:
General code cleanup

SarahPanther committed:
Changed ClimaCell v3 API url to tomorrow.io v4 - it seems no other changes are needed to our API call.  Close #19 ; Close #24

SarahPanther committed:
More code cleanup and commnets

SarahPanther committed:
Even more code cleanup -  comments and removing unused imports

SarahPanther committed:
Corrected an accidental change to the UI

SarahPanther committed:
Changed .gitignore and got rid of .idea folder

SarahPanther committed:
Got rid of useless dependencies and moved Level logic from MainActivity to Level

SarahPanther committed:
Added precipitationIntesity to updateWeatherData Dao call and Repository call.  Extracted Task key strings to TaskDataFinals class. Deleted onStopped overridden methods in GPSWorker and WeatherAPIWorker

SarahPanther committed:
Changed access modifiers for field stringsin CollectWeatherData and used context from constructor in WeatherAPIWorker

SarahPanther committed:
Merged sarah-merging-breathe-app-with-GPS-ClimaCell into merging-the-merging-branch-with-GPS-WeatherData-to-breathe-app

SarahPanther committed:
Refactored Workers for GPS and WeatherData getting and saving.  Can see log statements that the WeatherData saving is working but UI and file export don't show the new IUE's WeatherData

SarahPanther committed:
Fixed mistake in repository for passing the correct Instant into the workers. Fixed UI labels and string formatting, so WeatherData text views don't crash

SarahPanther committed:
Fixed some column errors in the csv file. Reformatted Level to have some conversion logic

Youssef-Beltagy committed:
removed old schema json

Youssef-Beltagy commented on Breathe/app/src/androidTest/java/com/ybeltagy/breathe/BreatheDaoTest.java:
```diff
@@ -162,18 +162,26 @@ public void updateWeatherDataTest(){
+        assertEquals(allEvents.get(0).getWeatherData().getWeatherGrassIndex(),
+                DataFinals.DEFAULT_LEVEL);
+        assertEquals(allEvents.get(0).getWeatherData().getWeatherEPAIndex(),
+                DataFinals.DEFAULT_INTEGER);
```
Add precipitation.

Youssef-Beltagy committed:
modified BreathDao

Youssef-Beltagy commented on Breathe/app/build.gradle:
```diff
@@ -40,6 +40,10 @@ android {
     implementation fileTree(dir: "libs", include: ["*.jar"])
 
+    // Dependencies needed for tomorrow.io API calls
+    implementation 'androidx.appcompat:appcompat:1.2.0'
```
not sure this is necessary.

Youssef-Beltagy commented on Breathe/app/build.gradle:
```diff
@@ -111,7 +116,16 @@ dependencies {
     implementation "androidx.work:work-multiprocess:$work_version"
 
+    // for serializing POJOs to be input and output from Tasks
+    implementation 'com.google.code.gson:gson:2.8.6'
```
should no longer be necessary after refactoring the serialization.

Youssef-Beltagy committed:
added precipitation in iue layout

Youssef-Beltagy committed:
work in progress

Youssef-Beltagy committed:
Resolved Conflicts
- Added support for weather data at all layers of the app.
- Fixed omission bugs
- Fixed some conflicts that cause bugs
- App works but the unit tests for tomorrow.io were not updated.
- Fix Bug: App attempts to get location without permission. Guard by checking for permission.

SarahPanther commented on Breathe/.gitignore:
```diff
@@ -12,3 +12,4 @@
 /captures
 .externalNativeBuild
 .cxx
+secrets.xml
```
Would it be better to put this in the shared preferences folder?

SarahPanther commented on Breathe/app/src/main/AndroidManifest.xml:
```diff
@@ -25,7 +25,11 @@
     <!--There is a good chance we need this permission. https://developer.android.com/reference/android/Manifest.permission#ACCESS_BACKGROUND_LOCATION
```
Deleting because it seems we don't need this (at least from what is needed in this branch)

SarahPanther committed:
Deleted the commented out permission for background location because we weren't using it

SarahPanther committed:
Removed placeholder IUE population of DB function and changed type converters to use Level's own conversion functions

SarahPanther commented on Breathe/app/src/main/java/com/ybeltagy/breathe/collection/WearableWorker.java:
```diff
@@ -42,13 +43,14 @@ public Result doWork() {
 
-        //todo: extract key
-        Instant timestamp = Instant.parse(getInputData().getString("timestamp"));
+        Instant timestamp = Instant.parse(getInputData().getString(TaskDataFinals.KEY_TIMESTAMP));
```
If the collection package is merged with the weather_data_collection package, this will no longer be referencing a final in another package. Leaving the package rearranging to the very last refactoring stage.

SarahPanther commented on Breathe/app/src/main/java/com/ybeltagy/breathe/data/BreatheDao.java:
```diff
@@ -107,18 +107,23 @@ int updateWearableData(Instant inhalerUsageTimeStamp, Instant wearableDataTimeSt
             "weatherHumidity = :weatherHumid, " +
-            "weatherPollen = :weatherPollen," +
-            "weatherAQI = :weatherAQI " +
+            "weatherPrecipitationIntensity = :weatherPrecipitation," +
```
Can we just pass in a WeatherData object? We need to test this.

SarahPanther committed:
Corrected a misspelled function name
